### PR TITLE
Jimix/23 pushpull

### LINKED ISSUE #23: Discuss the push/pull methods and how they should be revealed to the user
quote: @mkistler

> It seems that you have embedded the replicator functions into the incremental store. There is another approach that we should at least consider: exposing the local Cloudant-sync db and allow the application to manage replication.

### PR DESCRIPTION
**Pull request by jimix**
Resolves #23
@mkistler: please ACK when you get a chance

### CODE REVIEW

File: Classes/common/CDTIncrementalStore.h [hunk 1]
```diff
@@ -129,4 +59,34 @@ typedef NS_ENUM(NSInteger, CDTIncrementalStoreErrors) {
...
+ *  @return a CDTReplicator instance which can be used to start and
+ *  stop the replication itself, or `nil` on error.
+ */
+- (CDTReplicator *)replicatorThatPushesToURL:(NSURL *)remoteURL withError:(NSError **)error;
+
+/**
+ * Create a CDTReplicator object set up to replicate changes from a remote database to the
+ * local datastore.
+ *
+ *  @param remoteURL the remote server URL to which the data is replicated.
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.h:85]: "to which" should be "from which"

File: Classes/common/CDTIncrementalStore.m [hunk 2]
```diff
@@ -1571,75 +1370,73 @@ - (void)setMetadata:(NSDictionary *)metadata
...
-    CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: %@: state: %@", CDTISType,
-               [self cleanURL:self.remoteURL], replicator, state);
-}
+    CDTPushReplication *pushRep =
+        [CDTPushReplication replicationWithSource:self.datastore target:remoteURL];
+    if (!pushRep) {
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create push replication object",
+                    CDTISType, clean);
+        return nil;
+    }
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.m:1403]: Seems like we would not need to recreate the repFactory and pushRep every time this method is called.  We could cache these in local properties and reuse them.

File: doc/coredata.md [hunk 3]
```diff
@@ -99,49 +99,30 @@ NSURL *linkURL = [NSURL URLWithString:databaseURI];
...
-					        if (e) // ... deal with error
-						    [weakProgress setProgress:1.0 animated:YES];
-					    } else {
-					        [weakProgress setProgress:(float)processed / (float)total animated:YES];
-                        }
-                    }];
-if (!pull) // .. deal with error in `err`
-```
+In order to perform a push or a pull you need to obtain a
+`CDTReplicator` that can to the operation. These are provided by
```
Comment (mkistler) [inline doc/coredata.md]: "can to" -> "performs"

File: doc/coredata.md [hunk 4]
```diff
@@ -99,49 +99,30 @@ NSURL *linkURL = [NSURL URLWithString:databaseURI];
...
+`-replicatorThatPushesToURL:withError:` and
+`-replicatorThatPullsFromURL:withError:`. From this point you can
+assign use the replicator as instructed in [Replication], including
```
Comment (mkistler) [inline doc/coredata.md]: "assign use" -> "use"

File: doc/coredata.md [hunk 5]
```diff
@@ -99,49 +99,30 @@ NSURL *linkURL = [NSURL URLWithString:databaseURI];
...
+        [NSThread sleepForTimeInterval:1.0f];
     }
-}];
-
-if (!push) // .. deal with error in `err`
-
-while (!done) {
-    [NSThread sleepForTimeInterval:1.0f];
 }
 ```
```
Comment (mkistler) [inline doc/coredata.md:115]: I thought that there were some requirement on passing conflicts back into the CDTIncrementalStore so that these could be resolved.  Has that changed, or did I misunderstand.  If not, then we need to doc that here.

File: Classes/common/CDTIncrementalStore.h [hunk 6]
```diff
@@ -129,4 +59,34 @@ typedef NS_ENUM(NSInteger, CDTIncrementalStoreErrors) {
...
+ *  @param error     report error information
+ *
+ *  @return a CDTReplicator instance which can be used to start and
+ *  stop the replication itself, or `nil` on error.
+*/
+- (CDTReplicator *)replicatorThatPullsFromURL:(NSURL *)remoteURL withError:(NSError **)error;
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.h:91]: Since these replicatorThat... methods are instance methods on CDTIncrementalStore, the user will need to first get a reference to the proper CDTIncrementalStore object in order to use them.  I think that is non-obvious, so at a minimum we should give the user some guidance on how to do that, and maybe also consider extending the CDTIncrementalStore interface to make it easier.  E.g. we could offer a class method that takes a peristentStoreCoordinator and storeURL and returns a CDTIncrementalStore object.

File: Classes/common/CDTIncrementalStore.h (hunk 1 above)
Comment (jimix) [inline Classes/common/CDTIncrementalStore.h:85]: ack

File: Classes/common/CDTIncrementalStore.m (hunk 2 above)
Comment (jimix) [inline Classes/common/CDTIncrementalStore.m:1403]: I think it is possible that `repFactory` could be reused, I will check and I think we could just create it up front.
However, the `(push|pull)Rep` objects are actually single use, so we need to create them on demand.

File: doc/coredata.md (hunk 3 above)
Comment (jimix) [inline doc/coredata.md]: ACK

File: doc/coredata.md (hunk 4 above)
Comment (jimix) [inline doc/coredata.md]: ACK

File: doc/coredata.md (hunk 5 above)
Comment (jimix) [inline doc/coredata.md:115]: This patch is just getting rid of the old/broken push/pull methods.  Conflict resolution will be another PR

File: Classes/common/CDTIncrementalStore.h (hunk 6 above)
Comment (jimix) [inline Classes/common/CDTIncrementalStore.h:91]: This is covered in `## Accessing your store object`.
Having said that.. the discussion on `-linkReplicators:` needs to go.

File: Classes/common/CDTIncrementalStore.m (hunk 2 above)
Comment (jimix) [inline Classes/common/CDTIncrementalStore.m:1403]: I have confirmed that `repFactory` can be reused.

**Comment by jimix** [comment]
Well.. it turns out that the `CDTReplicatorFactory` is single use as well.
At least we have improved the error message :)

**Comment by mkistler** [comment]
ACK.  I've reviewed 425a7ed and all looks good to me.
